device_doctor/lib/src/android_device.dart

```diff
@@ -0,0 +1,195 @@
+
+  @visibleForTesting
+  AndroidDeviceDiscovery.testing();
+
+  // Parses information about a device. Example:
+  //
+  // 015d172c98400a03       device usb:340787200X product:nakasi model:Nexus_7 device:grouper
+  static final RegExp _kDeviceRegex = RegExp(r'^(\S+)\s+(\S+)(.*)');
+
+  static AndroidDeviceDiscovery _instance;
+
+  Future<String> deviceListOutput() async {
+    return eval(properties['adb'], <String>['devices', '-l'], canFail: false).timeout(Duration(seconds: 15));
+  }
+
+  Future<List<String>> deviceListOutputWithRetries(Duration retriesDelayMs) async {
+    RetryOptions r = RetryOptions(
+      maxAttempts: 3,
+      delayFactor: retriesDelayMs,
+    );
+    return await r.retry(
+      () async {
+        String result = await deviceListOutput();
+        return result.trim().split('\n');
+      },
+      retryIf: (Exception e) => e is TimeoutException,
+      //onRetry: (Exception e) => killAdbServer(),
+    );
+  }
+
+  void killAdbServer() async {
+    if (Platform.isWindows) {
+      await killAllRunningProcessesOnWindows('adb');
+    } else {
+      await eval(properties['adb'], <String>['kill-server'], canFail: false);
+    }
+  }
+
+  @override
+  Future<List<Device>> discoverDevices({Duration retriesDelayMs = const Duration(seconds: 10)}) async {
+    List<String> output = await deviceListOutputWithRetries(retriesDelayMs);
+    List<String> results = <String>[];
+    for (String line in output) {
+      // Skip lines like: * daemon started successfully *
+      if (line.startsWith('* daemon ')) continue;
+
+      if (line.startsWith('List of devices')) continue;
+
+      if (_kDeviceRegex.hasMatch(line)) {
+        Match match = _kDeviceRegex.firstMatch(line);
+
+        String deviceID = match[1];
+        String deviceState = match[2];
+
+        if (!const ['unauthorized', 'offline'].contains(deviceState)) {
+          results.add(deviceID);
+        }
+      } else {
+        throw 'Failed to parse device from adb output: $line';
+      }
+    }
+    return results.map((String id) => AndroidDevice(deviceId: id)).toList();
+  }
+
+  @override
+  Future<Map<String, List<HealthCheckResult>>> checkDevices() async {
+    final Map<String, List<HealthCheckResult>> results = <String, List<HealthCheckResult>>{};
+    for (AndroidDevice device in await discoverDevices()) {
+      final List<HealthCheckResult> checks = <HealthCheckResult>[];
+      checks.add(HealthCheckResult.success('device_access'));
+      HealthCheckResult check;
+      try {
+        // Just a smoke test that we can read wakefulness state
+        await device._getWakefulness();
+        check = await device.batteryHealth();
+      } catch (e, s) {
+        check = HealthCheckResult.error('battery_health', e, s);
+      }
+      checks.add(check);
+      results['android-device-${device.deviceId}'] = checks;
+    }
+    return results;
+  }
+
+  @override
+  Future<void> recoverDevices() async {
+    for (Device device in await discoverDevices()) {
```

Comment:
Should this be inherited from the parent class?

Reply:
the parent class is defined as an abstract class.

Reply:
Does it need to be abstract? If there's common code, we can put it there. Other methods can be required to be implemented.

Reply:
Yes. The class contains other functions as well which don't share common codes.